reconcile/external_resources/manager.py

```diff
@@ -0,0 +1,330 @@
+            resource_factories=setup_aws_resource_factories(
+                er_inventory, secret_reader
+            ),
+        ),
+    )
+    return of
+
+
+class ExternalResourcesManager:
+    def __init__(
+        self,
+        state_manager: ExternalResourcesStateDynamoDB,
+        settings: ExternalResourcesSettingsV1,
+        module_inventory: ModuleInventory,
+        reconciler: ExternalResourcesReconciler,
+        secret_reader: SecretReaderBase,
+        er_inventory: ExternalResourcesInventory,
+        factories: ObjectFactory[ExternalResourceFactory],
+        secrets_reconciler: InClusterSecretsReconciler,
+    ) -> None:
+        self.state_mgr = state_manager
+        self.settings = settings
+        self.module_inventory = module_inventory
+        self.reconciler = reconciler
+        self.er_inventory = er_inventory
+        self.factories = factories
+        self.secret_reader = secret_reader
+        self.secrets_reconciler = secrets_reconciler
+        self.errors: dict[ExternalResourceKey, ExternalResourceValidationError] = {}
+
+    def _resource_needs_reconciliation(
+        self,
+        reconciliation: Reconciliation,
+        state: ExternalResourceState,
+    ) -> bool:
+        reconcile: bool = False
+        reason: str = ""
+        if reconciliation.action == Action.APPLY:
+            match state.resource_status:
+                case ResourceStatus.NOT_EXISTS:
+                    reconcile = True
+                    reason = "Resource does not exist"
+                case ResourceStatus.ERROR:
+                    reconcile = True
+                    reason = "Resource status ERROR"
+                case ResourceStatus.CREATED:
+                    if reconciliation.resource_hash != state.resource_hash:
+                        reconcile = True
+                        reason = "Resource spec has changed"
+                    elif (
+                        (datetime.now(state.ts.tzinfo) - state.ts).total_seconds()
+                        > reconciliation.module_configuration.reconcile_drift_interval_minutes
+                        * 60
+                    ):
+                        reconcile = True
+                        reason = "Resource drift detection run"
+        elif reconciliation.action == Action.DESTROY:
+            match state.resource_status:
+                case ResourceStatus.CREATED:
+                    reconcile = True
+                    reason = "Resource needs to be deleted"
+                case ResourceStatus.ERROR:
+                    reconcile = True
+                    reason = "Resource status ERROR"
+        if reconcile:
+            logging.info(
+                "Reconciling: Status: [%s], Action: [%s], reason: [%s], key:[%s]",
+                state.resource_status.value,
+                reconciliation.action.value,
+                reason,
+                reconciliation.key,
+            )
+        return reconcile
```

Comment:
there are many `reconcile = True` and set `reason`, how about extract all reason string to enum, like `ReconcileStatus`, then it can be something like this:

```python
class ReconcileStatus(Enum):
    Reconciled = "Reconciled"
    NotExists = "NotExists"
    ...

    def _get_reconcile_status(
        self,
        reconciliation: Reconciliation,
        state: ExternalResourceState,
    ) -> ReconcileStatus:
        if reconciliation.action == Action.APPLY:
            match state.resource_status:
                case ResourceStatus.NOT_EXISTS:
                    return ReconcileStatus.NotExists
            ...
        if reconciliation.action == Action.DESTROY:
            ...

    def _resource_needs_reconciliation(
        self,
        reconciliation: Reconciliation,
        state: ExternalResourceState,
    ) -> bool:
        reconcile_status = self._get_reconcile_status(reconciliation, state)
        reconcile = reconcile_status != ReconcileStatus.Reconciled
        if reconcile:
            logging...
        return reconcile
```

Reply:
I changed the logic to something along the lines of what you have described. Thanks for the suggestion. PTAL